### PR TITLE
[react] add createFactory(type: ComponentType) definition

### PR DESCRIPTION
**Pull request by tpucci**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactjs.org/docs/react-api.html#createfactory
- [ ] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -209,7 +210,7 @@ declare namespace React {
         type: ClassType<P, ClassicComponent<P, ComponentState>, ClassicComponentClass<P>>): CFactory<P, ClassicComponent<P, ComponentState>>;
     function createFactory<P, T extends Component<P, ComponentState>, C extends ComponentClass<P>>(
         type: ClassType<P, T, C>): CFactory<P, T>;
-    function createFactory<P>(type: ComponentClass<P>): Factory<P>;
+    function createFactory<P>(type: ComponentType<P>): Factory<P>;
```
Comment (ferdaber) [inline types/react/index.d.ts]: I would actually use `JSXElementConstructor` instead, actually. It's more broad and doesn't require passing in function/class components that have the correct static properties).

Comment (tpucci) [reply types/react/index.d.ts]: Hi! I tried to replace with `ComponentClass` with `JSXElementConstructor`.
Then I encounter the following error:

```
ERROR: 254:7  expect  TypeScript@next compile error: 
Type 'Factory<SCProps>' is not assignable to type 'FunctionComponentFactory<SCProps>'.
  Type 'ReactElement<SCProps, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)> | null) | (new (props: any) => Component<any, any, any>)>' is not assignable to type 'FunctionComponentElement<SCProps>'.
    Types of property 'type' are incompatible.
      Type 'string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)> | null) | (new (props: any) => Component<any, any, any>)' is not assignable to type 'FunctionComponent<SCProps>'.
        Type 'string' is not assignable to type 'FunctionComponent<SCProps>'.
ERROR: 259:7  expect  TypeScript@next compile error: 
Type 'Factory<SCProps>' is not assignable to type 'FunctionComponentFactory<SCProps>'.
```

I am not sure how to solve this problem. Any guess ?

**Comment by RyanCavanaugh** [comment]
It'd be great to add a test so it's obvious to reviewers why this change is necessary

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]
We really need tests for any changes for a definition file this popular

**Comment by tpucci** [comment]
> We really need tests for any changes for a definition file this popular

I added a test. Hope everything is OK 😄 

Perhaps there is something I am missing but is there a way to efficiently test such type definition ?

I keep running `yarn test` at the root of the repo but it takes some times 🕐

**Pull request closed without merging**